### PR TITLE
Let user register callback and get acknowledgement result

### PR DESCRIPTION
**Pull request by NivedhaSenthil**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-cloud-ruby/issues) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea.
- [x] Follow the instructions in [CONTRIBUTING](CONTRIBUTING.md). Most importantly, ensure the tests and linter pass by running `bundle exec rake ci` in the gem subdirectory.
- [x] Update code documentation if necessary.

closes: #18236 
closes: #18237

### CODE REVIEW

**Comment by NivedhaSenthil** [comment]
This PR is in draft stage, to be reviewed once https://github.com/googleapis/google-cloud-ruby/pull/18395 is reviewed and merged.

File: google-cloud-pubsub/lib/google/cloud/pubsub/received_message.rb [hunk 1]
```diff
@@ -179,9 +181,9 @@ def published_at
         #   # Shut down the subscriber when ready to stop receiving messages.
         #   subscriber.stop!
         #
-        def acknowledge!
+        def acknowledge! &block
```
Comment (maheshgattani) [inline google-cloud-pubsub/lib/google/cloud/pubsub/received_message.rb:184]: This is my lack of ruby skills: Is this change impacting existing usage of acknowledge? Same question for the rest of the changes in this class.

I see we have changes in the usage comments. Do we need to provide example for previous usage?

File: google-cloud-pubsub/lib/google/cloud/pubsub/service.rb [hunk 2]
```diff
@@ -255,11 +255,7 @@ def streaming_pull request_enum
         ##
         # Acknowledges receipt of a message.
         def acknowledge subscription, *ack_ids
-          begin
```
Comment (maheshgattani) [inline google-cloud-pubsub/lib/google/cloud/pubsub/service.rb:258]: Why are we removing the try catch as part of this?

**Review by maheshgattani — COMMENTED** [review]
I have not looked at most of the PR. But I want to understand are we changing behavior or usage for non exactly once delivery subs? I expect no, but I am not skilled enough in Ruby to know that automatically.

File: google-cloud-pubsub/lib/google/cloud/pubsub/received_message.rb [hunk 3]
```diff
@@ -170,7 +170,9 @@ def published_at
         #   subscriber = sub.listen do |received_message|
         #     puts received_message.message.data
         #
-        #     received_message.acknowledge!
+        #     received_message.acknowledge! do |result|
+        #         puts result.status
```
Comment (dazuma) [inline google-cloud-pubsub/lib/google/cloud/pubsub/received_message.rb:174]: Please use the correct indentation for all sample code. (There are additional cases below.)

File: google-cloud-pubsub/lib/google/cloud/pubsub/received_message.rb (hunk 1 above)
Comment (dazuma) [reply google-cloud-pubsub/lib/google/cloud/pubsub/received_message.rb:184]: This is not a usage change. The `&` "argument" is just a syntax for capturing any passed block as an argument.

File: google-cloud-pubsub/lib/google/cloud/pubsub/acknowledge_result.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      ##
+      # The result of a ack/nack/modack on messages. The message object is available on
+      # {#message} and will have {#message_id} assigned by the API.
+      #
+      # When the operation was successful the result will be marked
+      # {#succeeded?}. Otherwise, the result will be marked {#failed?} and the
+      # error raised will be availabe on {#error}.
+      #
+      class AcknowledgeResult
+        SUCCESS = 1
```
Comment (dazuma) [inline google-cloud-pubsub/lib/google/cloud/pubsub/acknowledge_result.rb:28]: Please include documentation comments on all constants and attributes.

File: google-cloud-pubsub/lib/google/cloud/pubsub/acknowledge_result.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+        ##
+        # @private Create an PublishResult object.
+        def initialize status, error = nil
+          @error = error
+          @status = status
+        end
+
+        ##
+        # Whether the operation was successful.
```
Comment (dazuma) [inline google-cloud-pubsub/lib/google/cloud/pubsub/acknowledge_result.rb:45]: Include a `@return` type in the documentation. (Boolean in this case)

File: google-cloud-pubsub/lib/google/cloud/pubsub/received_message.rb (hunk 1 above)
Comment (dazuma) [inline google-cloud-pubsub/lib/google/cloud/pubsub/received_message.rb:184]: We do need to document what the block is for, when it is called, and what is passed to it.

**Review by dazuma — CHANGES_REQUESTED** [review]

File: google-cloud-pubsub/lib/google/cloud/pubsub/service.rb (hunk 2 above)
Comment (NivedhaSenthil) [reply google-cloud-pubsub/lib/google/cloud/pubsub/service.rb:258]: This try catch was added when we had to ignore all the GRPC errors earlier when we were yet to add support for exactly once delivery. As we will be now propagating the error and take action based on it like retry or constructing AcknowledgementResult we no longer need to catch it and ignore here.

**Review by dazuma — APPROVED** [review]

File: google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb [hunk 6]
```diff
@@ -53,31 +58,37 @@ def initialize subscriber, max_bytes: 500_000, interval: 1.0
             # entry.
             @register = {}
 
-            @retry_thread_pool = Concurrent::ThreadPoolExecutor.new max_threads: @subscriber.push_threads
+            @ack_callback_register = {}
+            @modack_callback_register = {}
+
+            @retry_thread_pool = Concurrent::ThreadPoolExecutor.new max_threads: @subscriber.callback_threads
```
Comment (Raou1d) [inline google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb:64]: Shouldn't this be `@subscriber.push_threads`?

**Pull request merged**